### PR TITLE
refactor Volume to Directory

### PR DESCRIPTION
**Pull request by TomasTomecek**

### CODE REVIEW

**Comment by psklenar** [comment]
## jenkins/superCI comment
* Status: success, return code = __0__ 
* Result url: __[https://da.gd/XY6TG](https://da.gd/XY6TG)__
* Comment: __This is testing CI__

File: conu/utils/core.py [hunk 1]
```diff
@@ -33,68 +35,87 @@ def random_str():
...
-    def set_force_selinux(self, value):
-        self.force_selinux = value
-
-    def set_permission(self, perms):
-        run_cmd(["chmod", perms, self.directory])
-
-    def set_facl(self, rules):
-        for rule in rules:
-            run_cmd(["setfacl", "-m", rule, self.directory])
+class Directory(object):
```
Comment (dhodovsk) [inline conu/utils/core.py]: I'd suggest moving this to another module and using core(maybe renamed) for little non-class utilities.

File: conu/utils/core.py [hunk 2]
```diff
@@ -6,25 +6,27 @@
...
-    if not raw:
-        logger.debug("command: %s" % cmd)
-        if isinstance(cmd, str):
-            output = subprocess.check_output(cmd.split(" "), **kwargs)
-        else:
-            output = subprocess.check_output(cmd, **kwargs)
+# TODO: utilize this function everywhere and replace subprocess usage
+def run_cmd(cmd, return_output=False):
+    """
+    run provided command, raise subprocess.CalledProcessError if it fails
```
Comment (dhodovsk) [inline conu/utils/core.py]: Here it could be emphasized that it is being run on host and under which user.

File: conu/utils/core.py [hunk 3]
```diff
@@ -33,68 +35,87 @@ def random_str():
...
+            self._initialized = True
+            logger.info("initialized")
+            return
+        logger.info("%r was already initialized", self)
+
+    def _set_selinux_context(self):
+        # FIXME: do this using python API if possible
+        if self.context:
+            logger.debug("setting SELinux context of %s to %s", self.path, self.context)
+            run_cmd(["chcon", "-t", self.context, self.path])
```
Comment (dhodovsk) [inline conu/utils/core.py]: With -t switch only selinux type is changed. (As I understood selinux context = user:role:type:level) Name of function should be then set_selinux_type (or other should be added).

File: conu/utils/core.py [hunk 4]
```diff
@@ -33,68 +35,87 @@ def random_str():
...
+
+    def _set_mode(self):
+        if self.mode is not None:
+            logger.debug("changing permission bits of %s to %s", self.path, self.mode)
```
Comment (dhodovsk) [inline conu/utils/core.py]: Print mode in oct

File: conu/utils/core.py [hunk 5]
```diff
@@ -33,68 +35,87 @@ def random_str():
...
+    def __enter__(self):
+        self.initialize()
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.clean()
 
     def clean(self):
-        shutil.rmtree(self.directory)
+        if self._initialized:
```
Comment (dhodovsk) [inline conu/utils/core.py]: docstring missing :(

File: conu/utils/core.py [hunk 6]
```diff
@@ -33,68 +35,87 @@ def random_str():
...
+                    os.makedirs(self.path)
+            self._set_mode()
+            self._add_facl_rules()
+            self._set_selinux_context()
+            self._initialized = True
+            logger.info("initialized")
+            return
+        logger.info("%r was already initialized", self)
+
+    def _set_selinux_context(self):
```
Comment (dhodovsk) [inline conu/utils/core.py]: This bit fails on systems with SELinux disabled. Maybe there could be check (sestatus) and warning that context won't be set when selinux is disabled.

File: conu/utils/core.py (hunk 2 above)
Comment (TomasTomecek) [reply conu/utils/core.py]: will do

File: conu/utils/core.py (hunk 1 above)
Comment (TomasTomecek) [reply conu/utils/core.py]: +1

File: conu/utils/core.py (hunk 5 above)
Comment (TomasTomecek) [reply conu/utils/core.py]: you're good!

File: conu/utils/core.py (hunk 6 above)
Comment (TomasTomecek) [reply conu/utils/core.py]: very good point!

**Comment by TomasTomecek** [comment]
I hope I addressed all of them, let's see what CI will tell, b/c I was getting some odd failures locally

**Comment by psklenar** [comment]
## jenkins/superCI comment
* Status: success, return code = __0__ 
* Result url: __[https://da.gd/h7L5](https://da.gd/h7L5)__
* Comment: __This is testing CI__

**Comment by psklenar** [comment]
## jenkins/superCI comment
* Status: failure, return code = __1__ 
* Result url: __[https://da.gd/VLtRG](https://da.gd/VLtRG)__
* Comment: __This is testing CI__

**Comment by psklenar** [comment]
## jenkins/superCI comment
* Status: success, return code = __0__ 
* Result url: __[https://da.gd/IcmPk](https://da.gd/IcmPk)__
* Comment: __This is testing CI__

**Pull request merged**